packages/runtime/src/logicsig.ts

```diff
@@ -26,10 +27,22 @@ export class LogicSig {
   lsigAddress: string;
   tag: Buffer;
 
-  constructor (program: string, programArgs?: Array<Uint8Array | Buffer> | null) {
+  constructor (
+    logicSigAddresses: Map<number, string>, program: string,
+    programArgs?: Array<Uint8Array | Buffer> | null
+  ) {
     this.tag = Buffer.from("Program");
     this.logic = parsing.stringToBytes(program);
-    this.lsigAddress = generateAccount().addr;
+
+    const hash = murmurhash.v3(program);
+    const result = logicSigAddresses.get(hash);
+    if (result === undefined) {
+      const addr = generateAccount().addr;
```

Comment:
how about the following:
* `generateAccount` will take an optional argument: `code: string`.
* if `code` is provided it will use it to generate an address by hashing the code.
* otherwise it will use a random number.

This way:
* we don't need `logicSigAddresses` map
* we assure the addresses will be deterministic for lsig.

Reply:
i thought this, i wanted to keep same address as algorand but to do so, we need a logic of program in bytes which is returned by compiler and then uint8array is used to create a hash like this:
```
const toBeSigned = utils.concatArrays(this.tag, this.logic);
const hash = nacl.genericHash(toBeSigned);
return address.encodeAddress(new Uint8Array(hash));
```
But we cannot get uint8array that's why i didn't follow this.
will that be ok if address is not `ALGORAND ADDRESS` ?
looking more into encoding algorand addresses

Reply:
![Screenshot from 2021-09-16 11-50-39](https://user-images.githubusercontent.com/26249903/133561326-54feed86-b476-4579-9304-d37e9efa5b9e.png)

I tried to use hash to make `encodeAddress(parsing.uint64ToBigEndian(hash));`

also i tried using `encodeAddress(parsing.stringToBytes(program));`
![as](https://user-images.githubusercontent.com/26249903/133561620-ab705db5-7923-4490-9923-7414f45f2039.png)

Reply:
verification is failing because address is not `ALGORAND_ADDRESS`, we have added test for verification so that only logic sig address can send transactions from logic sig.

Reply:
found js-sdk is sending request to `.compile(),do()` to go-algorand and then there we have compiler which is used to compile string to bytes

Reply:
We just need to find how to construct a valid algorand address based on a seed (here the hash of the code).

Here is how the address is constructed: https://developer.algorand.org/docs/features/accounts/#transformation-public-key-to-algorand-address

So what we need:
1. `key = hash(code).slice(0, 32)`  -- take 32 bytes from the code hash
2. `checksum = sha256(key)`
3. `address_bytes = key.concat(checksum.slice(-4))`
4. `address = base32(address_bytes)`

Reply:
done added key_pair generation using random seed, random seed used is hash of program